Add support for Playlists, etc (#656)

skiphansen opened the pull request:
This Wiki [page](https://github.com/skiphansen/pianobar/wiki/Premium-Features) describes these changes from the user's viewpoint.

I use pianobar for playback but I use the Pandora App to manage my "collection" so I have not added any support for **creating or editing** playlists or collecting tracks.

I have however added support for deleting podcasts, albums, playlists and tracks to the delete command.

Most of the new changes are only of any value to those with a premium account, If it's not worth your effort to review/merge the changes I understand.

skiphansen committed:
Added support for building using specified location for ffmpeg libs by setting LIBAV.

skiphansen committed:
Merge branch 'master' of https://github.com/PromyLOPh/pianobar

skiphansen committed:
Added support for podcasts, playlists, albums and tracks (#656).

PromyLOPh commented on Makefile:
```diff
@@ -41,13 +46,20 @@ LIBPIANO_SRC:=\
 
-LIBAV_CFLAGS:=$(shell $(PKG_CONFIG) --cflags libavcodec libavformat libavutil libavfilter)
-LIBAV_LDFLAGS:=$(shell $(PKG_CONFIG) --libs libavcodec libavformat libavutil libavfilter)
+LIBAV = /home/skip/open_src/audio/ffmpeg
```
Probably a leftover from your own debugging.

skiphansen replied:
Sort of, my ffmpeg library wasn't compatible so I built a local copy.  Clearly this shouldn't be part of the PR.  I'll fix it.

PromyLOPh commented on src/libpiano/request.c:
```diff
@@ -501,8 +505,158 @@ PianoReturn_t PianoRequest (PianoHandle_t *ph, PianoRequest_t *req,
+			json_object_object_add(a,"limit",json_object_new_int(100));
+			json_object_object_add(a,"annotationLimit",json_object_new_int(100));
+			json_object_object_add(j,"request",a);
+			json_object_object_add(j,"deviceId",json_object_new_string("1880"));
```
This looks wrong. The device id is probably specific to a user, no?

skiphansen replied:
I don't believe so I think it's analogous to devceid in [partner passwords](https://6xq.net/pandora-apidoc/json/partners/ ).  Since this comes from reverse engineering and sniffing I certainly could be wrong.  In any case if I'll try to check if it changes with different user accounts.

skiphansen replied:
I removed the deviceId parameter from all requests and everything still works.

PromyLOPh commented on src/libpiano/request.c:
```diff
@@ -501,8 +505,158 @@ PianoReturn_t PianoRequest (PianoHandle_t *ph, PianoRequest_t *req,
+				case PIANO_TYPE_PLAYLIST: {
+					json_object *a = json_object_new_object ();
+					json_object_object_add(a,"pandoraId",json_object_new_string(reqData->station->id));
+					json_object_object_add(a,"limit",json_object_new_int(100));
```
We’d probably need a very large number here or some kind of pagingation aka `PIANO_RET_CONTINUE_REQUEST`.

PromyLOPh commented on src/libpiano/request.c:
```diff
@@ -501,8 +505,158 @@ PianoReturn_t PianoRequest (PianoHandle_t *ph, PianoRequest_t *req,
+					json_object_object_add(a,"pandoraId",json_object_new_string(reqData->station->id));
+					json_object_object_add(a,"limit",json_object_new_int(100));
+					json_object_object_add(a,"annotationLimit",json_object_new_int(100));
+					json_object_object_add(a,"bypassPrivacyRu1les",json_object_new_boolean(true));
```
`Ru1les` looks like a typo.

skiphansen replied:
Yes it does, but by a Pandora developer not me!

I got that from sniffing the traffic.

```json
Request to playlists.v7.getTracks:
{
    "request": {
        "pandoraId": "PL:x:x",
        "limit": 100,
        "annotationLimit": 100,
        "bypassPrivacyRu1les": true
    },
    "deviceId": "1880",
    "userAuthToken": "x",
    "syncTime": 1729893293
}
```

PromyLOPh commented on src/libpiano/request.c:
```diff
@@ -501,8 +505,158 @@ PianoReturn_t PianoRequest (PianoHandle_t *ph, PianoRequest_t *req,
+			json_object *a = json_object_new_object ();
+			json_object_object_add(j,"request",a);
+	#if 0
+	// test ability to continue after hitting limit
```
So, we also need a continue return value here?

PromyLOPh commented on src/libpiano/response.c:
```diff
@@ -241,6 +316,173 @@ PianoReturn_t PianoResponse (PianoHandle_t *ph, PianoRequest_t *req) {
+						song->artist = PianoJsonStrdup(trackInfo, "artistName");
+						song->album = PianoJsonStrdup(trackInfo, "albumName");
+						song->title = PianoJsonStrdup(trackInfo, "name");
+						song->fileGain = 0.0;
```
This does not exist in the response?

skiphansen replied:
No it isn't.  This is just to ensure the value is initialized to something sane in case it's not set elsewhere.

PromyLOPh commented on src/libpiano/response.c:
```diff
@@ -241,6 +316,173 @@ PianoReturn_t PianoResponse (PianoHandle_t *ph, PianoRequest_t *req) {
+							continue;
+						}
+						LOG("got track %s: ",Key);
+						trackNumber = getInt(Val,"trackNumber");
```
It this is the album track, we could probably extend `PianoSong_t` and add a `albumTrackNumber` field, so we don’t have to hard-code this here. When listing an album with `G`, it’ll be numbered anyways.

PromyLOPh commented on src/libpiano/response.c:
```diff
@@ -241,6 +316,173 @@ PianoReturn_t PianoResponse (PianoHandle_t *ph, PianoRequest_t *req) {
+						if(playlist == NULL) {
+							playlist = song;
+						}
+						else {
+							PianoSong_t *lastSong = (PianoSong_t *) &playlist;
+							PianoSong_t *nextSong = playlist;
+							do {
+								int nextSongTrackNum;
+								if(nextSong == NULL) {
+									lastSong->head.next = (struct PianoListHead *) song;
+									break;
+								}
+								sscanf(nextSong->title,"%d",&nextSongTrackNum);
+								if(nextSongTrackNum > trackNumber) {
+									lastSong->head.next = (struct PianoListHead *) song;
+									song->head.next = (struct PianoListHead *) nextSong;
+									break;
+								}
+								lastSong = nextSong;
+								nextSong = (PianoSong_t *) nextSong->head.next;
+							} while(true);
```
So, basically, the result returned by Pandora is not sorted and we need to do that here?

PromyLOPh commented on src/libpiano/response.c:
```diff
@@ -712,8 +955,370 @@ PianoReturn_t PianoResponse (PianoHandle_t *ph, PianoRequest_t *req) {
+				const char *type = PianoJsonGetStr(Val,"type");
+
+				Annotations++;
+				if(strcmp(type,"PL") == 0) {
+					ph->user.PlayListCount++;
+				}
+				else if(strcmp(type,"ST") == 0) {
+					ph->user.StationCount++;
+				}
+				else if(strcmp(type,"AL") == 0) {
+					ph->user.AlbumCount++;
+				}
+				else if(strcmp(type,"TR") == 0) {
+					ph->user.TrackCount++;
+				}
+				else if(strcmp(type,"LI") == 0) {
+				}
+				else if(strcmp(type,"AR") == 0) {
+				}
+				else {
+					LOG("type %s ignored\n",type);
+				}
```
Do we actually use these values somewhere? Do we need to store them? pianobar uses linked lists for everything anyways.

PromyLOPh commented on src/libpiano/response.c:
```diff
@@ -712,8 +955,370 @@ PianoReturn_t PianoResponse (PianoHandle_t *ph, PianoRequest_t *req) {
+				PianoStationType_t stationType = PIANO_TYPE_NONE;
+
+				if(strcmp(type,"PL") == 0 || strcmp(type,"ST") == 0) {
+				// Playlists and stations handled elsewhere
```
assert?

skiphansen replied:
Good idea, I'll add one.

PromyLOPh commented on src/libpiano/response.c:
```diff
@@ -712,8 +955,370 @@ PianoReturn_t PianoResponse (PianoHandle_t *ph, PianoRequest_t *req) {
+				song->trackToken = strdup(trackToken);
+				song->length = getInt(Val, "duration");
+			// Save release date for sorting
+				song->fileGain = ((Year - 1900) * 10000) + (Month * 100) + Day;
```
Just add a new field to `PianoSong_t` instead of misusing the gain.

skiphansen replied:
Will do.

PromyLOPh commented on src/main.c:
```diff
@@ -178,6 +179,53 @@ static bool BarMainGetStations (BarApp_t *app) {
+		if(!ret) {
+			break;
+		}
+		if(app->ph.user.IsPremiumUser) {
```
I’d remove these kind of checks unless the API calls always return an error for non-premium users (apparently they don’t).

PromyLOPh commented on src/ui.c:
```diff
@@ -887,6 +942,37 @@ static void BarUiEventcmdPrintSong (FILE * restrict stream,
+ * @param auto-select if only one station remains after filtering
+ * @return pointer to selected station or NULL
+ */
+void BarUiSelectFilter(BarApp_t *app)
```
I don’t like this UI, to be honest. It requires me to go to a different menu to filter the station selection by type. The idea of station selection was that you can just type name/id of the station and filter through that. It’d be more fitting if we could add some operator(?) to `BarUiSelectStation`.

PromyLOPh commented on src/ui_act.c:
```diff
@@ -938,3 +983,130 @@ BarUiActCallback(BarUiActManageStation) {
+			break;
+		}
+
+		case PIANO_TYPE_STATION:
```
We could easily implement this for ordinary stations.

PromyLOPh reviewed (commented):
Sorry for the months of silence :( I finally had a look at the code and tried your branch. My problem as a maintainer is that your changes are pretty invasive and I cannot test alot of them myself due to the lack of a premium subscription (and not living inside the US). I was hoping that – on the pianobar UI side – we would not have to differentiate between stations, playlists, podcasts, … and that we could hide this complexity in libpiano. I’m going to add a few review comments, but I’m unsure how to proceed.

PromyLOPh commented:
As another thought: We already have a song “history” and “upcoming“ songs. If we add support for playlists, we might as well merge all of that into a single, big “playlist“ functionality, allowing to move freely between past and future songs (only if we can fetch a new song URL, obviously).

skiphansen commented:
> Sorry for the months of silence :( I finally had a look at the code and tried your branch. My problem as a maintainer is that your changes are pretty invasive and I cannot test alot of them myself due to the lack of a premium subscription (and not living inside the US). I was hoping that – on the pianobar UI side – we would not have to differentiate between stations, playlists, podcasts, … and that we could hide this complexity in libpiano. I’m going to add a few review comments, but I’m unsure how to proceed.

Thanks for the feedback!  

I was surprised by how invasive the changes became, but once I got the hang of reverse engineering the protocol I got a bit carried away.  

I'm an old fart who typically plays "albums" from beginning to end while focused on listening, otherwise the "station" concept works for me for background music. Personally I'm not that interested in playlists, tracks or podcasts, but since they were there ...

I'll look into the detailed code comments in the next few days.

skiphansen commented:
> As another thought: We already have a song “history” and “upcoming“ songs. If we add support for playlists, we might as well merge all of that into a single, big “playlist“ functionality, allowing to move freely between past and future songs (only if we can fetch a new song URL, obviously).

Yes, but I don't think you can fetch a new song URL with a non-paid account.  I'm not sure, I haven't done much testing with a free account.

skiphansen committed:
Removed LIBAV init.

skiphansen committed:
Verify that pandoraType is found in GetItems response (review comment).

skiphansen committed:
Added releaseDate field to PianoSong_t instead of misusing fileGain.

skiphansen commented:
> Sorry for the months of silence :( I finally had a look at the code and tried your branch. My problem as a maintainer is that your changes are pretty invasive and I cannot test alot of them myself due to the lack of a premium subscription (and not living inside the US). I was hoping that – on the pianobar UI side – we would not have to differentiate between stations, playlists, podcasts, … and that we could hide this complexity in libpiano. I’m going to add a few review comments, but I’m unsure how to proceed.

I've addressed several of your comments and updated the PR.

The rest will require some work which I'm fine with if our eventual goal is to merge this PR.  

As far as testing goes is it possible for you to create a throwaway account, start a premium subscription trial and then cancel before the end of the trial period?

If it's just not worth your effort and/or not practical I'm fine with that but I probably won't make additional changes since the current code is filling my needs.

skiphansen committed:
Revert unintentional formating changes (tabs -> spaces).

skiphansen committed:
Removed deviceId parameter from getTracks, getItems and getSortedPlaylists requests.
Apparenly not needed and meaning is unknown.
